hudi-common/src/main/java/org/apache/hudi/common/model/HoodieFileGroup.java

```diff
@@ -122,7 +129,24 @@ public HoodieFileGroupId getFileGroupId() {
    * some log files, that are based off a commit or delta commit.
    */
   private boolean isFileSliceCommitted(FileSlice slice) {
-    if (!compareTimestamps(slice.getBaseInstantTime(), LESSER_THAN_OR_EQUALS, lastInstant.get().getTimestamp())) {
+    if (compareTimestamps(slice.getBaseInstantTime(), GREATER_THAN, lastInstant.get().getTimestamp())) {
+      return false;
+    }
+
+    if (!slice.getBaseFile().isPresent() && timeline.isBeforeTimelineStarts(slice.getBaseInstantTime())) {
```

Comment:
@voonhous : Not sure I am following this. when rollback happens, command block gets added to skip rollbacked log blocks.

Reply:
@bvaradar Reverted the change, please take a look. 

Understood. The crux of the issue that I am trying to fix is somewhat similar to a multi-writer scenario. 

In Flink, the job manager (JM) is responsible for performing a rollback, while the task manager (TM) is responsible for performing writes. 

It is entirely possible and very common for TM to perform writes before a JM performs a rollback (see https://issues.apache.org/jira/browse/HUDI-5822) when a job is recovering from and is restarting.

Under the bucket index use-case, a bucketId can only have 1 fileGroup. Using `getLatestFileSlices` when JM has yet to complete a rollback will cause the fileGroup that is pending rollback completion to not be visible to the TM.

TM will hence generate a new fileGroup for the same bucketId. Once the rollback completes, a partition might have a bucketId that maps to two fileGroups, breaking the 1 bucketId <> 1 fileGroup mapping contract.

As such, this was what #5185 was trying to fix. And allow fileGroups pending rollback to be re-used when performing a bucketIndex bootstrap. 

I found that this fix was a tad hacky, which was why I tried modifying the lower level APIs to address such scenarios, which as can be seen, i shortly gave up due to the performance penalties that you have highlighted.

Reply:
@voonhous : thanks for the great explanation. From your comment, it looks like there is a difference  in the way Flink integration resolves valid files. 

Regarding your comment: 
```Once the rollback completes, a partition might have a bucketId that maps to two fileGroups, breaking the 1 bucketId <> 1 fileGroup mapping contract.```

As part of rollback, shouldn't the underlying file (which was newly created as part of failed commit gettting rolled back) get deleted ? Also, why does the fileId getting read if the commit did not finish ?

Reply:
After reading your comment in [BucketStreamWriteFunction](https://github.com/apache/hudi/pull/7997/files#diff-bc2aaf0958acca5adb72c26a76348ab70c994929ac76a7befe70a263ffb66353), I understand the difference in COW and MOR behavior. The cleaner way would be to have MOR rollback logic delete the log files if the commit to be rollback created that file-group.  This avoids special casing in other places.

Reply:
IIUC, this issue is only specific to Flink writers as the the write and rollback logic might execute concurrently. 

While this will fix the issue, i agree that the cleaner way is for MOR to delete the log files so that the responsibility of writers and rollback executors are clearly defined. i.e. Writers should not know or handle the quirks of rollback executors. The current way of fixing things does introduce code smell.

To be honest, I am getting a lot of resistance and pushback with this fix, so I am going to take small steps. I have already spent 8 hours writing a test case and there doesn't seem to be an agreement in which direction everybody wants to take. 

@danny0405 @bvaradar LMK which direction the both of you would like to take to fix this issue. 

Thank you.

Reply:
I don't think this is a Flink specific bug, agree with @bvaradar somwhow the mor log rollback or the fs view(maybe related with metadata) should be fixed, let's create another JIRA issue to track the improvement.

To eradicate the bug from the root cause takes some time, I also agree @voonhous we can have a quick fix on Flink side.

@voonhous The PT overall looks good, can we move the tests into `ITTestDataStreamWrite` to reuse some tool methods, or can we re-organize these two test cases for some reuse things, because there are alreay bucket index tests in `ITTestDataStreamWrite`.

Reply:
@bvaradar @danny0405 there seems to be some discrepancy the 2 rollback strategies below:

1. rollback with marker - will NOT delete log files that were created by the commit to rollback
2. rollback via listing - will delete log files that were created by the commit to rollback

``` java
    // Log file can be deleted if the commit to rollback is also the commit that created the fileGroup
    if (latestLogFileOption.isPresent() && Objects.equals(baseCommitTime, instantToRollback.getTimestamp())) {
      Path fullDeletePath = new Path(partitionPath, latestLogFileOption.get().getFileName());
      return new HoodieRollbackRequest(relativePartitionPath, EMPTY_STRING, EMPTY_STRING,
          Collections.singletonList(fullDeletePath.toString()),
          Collections.emptyMap());
    }
```

Perhaps, a fix like this would be sufficient to resolve the discrepancy between how these 2 strategies rollback log files?

Reply:
Yeah, let's move to that direction.

Reply:
+1

Reply:
Please take a look at the fixes again. 

I don't think it's possible to move the test to `ITTestDataStreamWrite` as the tests assume that the end-state of the table will only have parquet files. 

The same assumption cannot be made for the tests that i added in `ITTestBucketStreamWrite`.

Reply:
I mean to fix the log file cleaning of when marker file rollback is enabled(by default), BTW, I didn't see the code snippet you pasted, can you show the permenent link instead?